### PR TITLE
Store digested dictionary in BlockBasedTableBuilder

### PR DESCRIPTION
**Pull request by ajkr**
We use the same compression dictionary to compress every data block in a given SST file. Previously, compressing each data block involved some repetitive work to convert the raw compression dictionary bytes into a usable form.

ZSTD's bulk processing API lets us avoid this repetitive work by providing APIs that create a "digested dictionary" (`ZSTD_createCDict`) and a compression function that operates on digested dictionaries (`ZSTD_compress_usingCDict`). This PR changes `BlockBasedTableBuilder` to create a single digested dictionary that is reused for compressing every data block. It is supported for zstd v0.7.0 and later.

Test Plan: verified it's faster; `make check -j64`; will do zstd version compatibility testing 

- Populate DB command:

```
TEST_TMPDIR=/dev/shm ./db_bench -benchmarks=filluniquerandom,compact -num=20000000 -compression_type=zstd -max_background_jobs=24 -memtablerep=vector -allow_concurrent_memtable_write=false -disable_wal=true -max_write_buffer_number=8
```

- Benchmark command:

```
TEST_TMPDIR=/dev/shm /usr/bin/time ./db_bench -use_existing_db=true -benchmarks=compact -compression_type=zstd -block_size=4096 -compression_max_dict_bytes=$dict_bytes
```

- Results before this PR:
```
dict_bytes  cpu_seconds
0           32.17
4096        41.41
16384       84.3
```

- Results after this PR:
```
dict_bytes  cpu_seconds
0           32.09
4096        38.51
16384       50.36
```

### CODE REVIEW

File: util/compression.h [hunk 1]
```diff
@@ -46,6 +46,75 @@
...
+#endif
+  Slice raw_dict_bytes;
+
+  DictContext() {}
+
+  DictContext(DictContext&& other) {
+    *this = std::move(other);
+  }
+
+  DictContext& operator=(DictContext&& other) {
```
Comment (siying) [inline util/compression.h]: Where do we need to make a copy?

Comment (ajkr) [reply util/compression.h]: Oh it's move-assignment, which is just used by the move-constructor above. Now that I rethink it, we could probably just call `Init` in the `Rep` constructor and never have to move the `DictContext` anywhere...

Comment (yuslepukhin) [reply util/compression.h]: My choice would be to explicitly invoke default __Ctor in the init for the move __ctor 
DictContext(DictContext&& other) : DictContext()

**Comment by ajkr** [comment]
replaced by #4251.

**Pull request closed without merging**